### PR TITLE
Batt smbus rename to bq40z50, ceate battery driver folder, and update bq40z50 driver.

### PR DESCRIPTION
**Pull request by AlexKlimaj**
This PR is to rename the batt_smbus driver to bq40z50 in order to reflect accurately what it is for. This also creates a battery folder in drivers for it to reside. This brings it in line with the rest of the drivers. 

It also adds two fields to battery_status.msg. A voltage array for individual cell voltages, and a cell voltage delta value. This is to communicate the max difference between the highest and lowest voltage cell in a pack. This can help diagnose bad batteries with imbalanced cells. 

mavlink_messages.cpp is also updated to populate the voltage array, and add time_remaining. 

```
INFO  [bq40z50] starting /dev/bq40z50_ext
BQ40Z50_I2C on I2C bus 1 at 0x0b (bus: 100 KHz, max: 100 KHz)
INFO  [bq40z50] Smart battery found at 0xb
INFO  [bq40z50] Smart battery connected
INFO  [bq40z50] bad PEC from block_read
INFO  [bq40z50] Max Cell Delta: 0.12

WARN  [bq40z50] Please enter an appropriate command.

Usage: bq40z50 <command> [arguments...]
 Commands:

   start
     [-X <val>]  (null)
                 default: BQ40Z50_BUS_I2C_EXTERNAL
     [-I <val>]  (null)
                 default: BQ40Z50_BUS_I2C_INTERNAL
     [-A <val>]  (null)
                 default: BQ40Z50_BUS_ALL

   stop          Stops the driver.

   suspend       Suspends the drive but does stop it completely.

   resume        Resumes the driver from the suspended state.

   man_nam       Prints the name of the manufacturer.

   man_date      Prints the date of manufacture.

   serial_num    Prints the serial number.

   sbs_info      Prints the manufacturer name, date, and serial number.

   info          Prints the last report.

   unseal        Unseals the devices flash memory to enable write_flash commands.

   read_word     Uses the SMbus read-word command.
     [command code] The SMbus command .

   man_read      Uses the SMbus block-read with ManufacturerAccess().
     [command code] The SMbus command .
     [number of bytes] Number of bytes to read.

   read_flash    Reads 32 bytes from flash starting from the address specified.
     [address  ] The address to start reading from. .

   write_flash   Writes to flash. The device must first be unsealed with the unseal command.
     [address  ] The address to start writing.
     [number of bytes] Number of bytes to send.
     [data[0]...data[n]] One byte of data at a time separated by spaces.

   block_read    Performs a SMBus block read.
     [command code] The SMbus command .
     [number of bytes] Number of bytes to read.

   block_write   Performs a SMBus block write.
     [command code] The SMbus command code.
     [number of bytes] Number of bytes to send.
     [data[0]...data[n]] One byte of data at a time separated by spaces.
nsh>
nsh>
nsh> bq40z50
WARN  [bq40z50] Please enter an appropriate command.

Usage: bq40z50 <command> [arguments...]
 Commands:

   start
     [-X <val>]  (null)
                 default: BQ40Z50_BUS_I2C_EXTERNAL
     [-I <val>]  (null)
                 default: BQ40Z50_BUS_I2C_INTERNAL
     [-A <val>]  (null)
                 default: BQ40Z50_BUS_ALL

   stop          Stops the driver.

   suspend       Suspends the drive but does stop it completely.

   resume        Resumes the driver from the suspended state.

   man_nam       Prints the name of the manufacturer.

   man_date      Prints the date of manufacture.

   serial_num    Prints the serial number.

   sbs_info      Prints the manufacturer name, date, and serial number.

   info          Prints the last report.

   unseal        Unseals the devices flash memory to enable write_flash commands.

   read_word     Uses the SMbus read-word command.
     [command code] The SMbus command .

   man_read      Uses the SMbus block-read with ManufacturerAccess().
     [command code] The SMbus command .
     [number of bytes] Number of bytes to read.

   read_flash    Reads 32 bytes from flash starting from the address specified.
     [address  ] The address to start reading from. .

   write_flash   Writes to flash. The device must first be unsealed with the unseal command.
     [address  ] The address to start writing.
     [number of bytes] Number of bytes to send.
     [data[0]...data[n]] One byte of data at a time separated by spaces.

   block_read    Performs a SMBus block read.
     [command code] The SMbus command .
     [number of bytes] Number of bytes to read.

   block_write   Performs a SMBus block write.
     [command code] The SMbus command code.
     [number of bytes] Number of bytes to send.
     [data[0]...data[n]] One byte of data at a time separated by spaces.
nsh> bq40z50 info
 battery_status_s
        timestamp: 27556283  (0.084550 seconds ago)
        voltage_v: 16.551
        voltage_filtered_v: 16.551
        current_a: 0.752
        current_filtered_a: 0.752
        average_current_a: 0.725
        discharged_mah: 4.000
        remaining: 0.950
        scale: 0.000
        temperature: 21.750
        cell_count: 4
        voltage_cell_v: [4.135, 4.139, 4.137, 4.139]
        max_cell_voltage_delta: 0.004
        capacity: 2604
        cycle_count: 4
        run_time_to_empty: 199
        average_time_to_empty: 205
        serial_number: 307
        connected: True
        system_source: False
        priority: 0
        is_powering_off: False
        warning: 0
nsh>
```

### CODE REVIEW

**Comment by dagar** [comment]
Why does this device need command line helpers for reading random memory?

File: src/modules/mavlink/mavlink_messages.cpp [hunk 1]
```diff
@@ -597,7 +598,7 @@ class MavlinkStreamSysStatus : public MavlinkStream
 
 			for (unsigned int i = 0; i < (sizeof(bat_msg.voltages) / sizeof(bat_msg.voltages[0])); i++) {
 				if ((int)i < battery_status.cell_count && battery_status.connected) {
-					bat_msg.voltages[i] = (battery_status.voltage_v / battery_status.cell_count) * 1000.0f;
+					bat_msg.voltages[i] = battery_status.voltage_cell_v[i] * 1000.0f;
```
Comment (LorenzMeier) [inline src/modules/mavlink/mavlink_messages.cpp]: This assumes that cell voltages are valid.

File: cmake/configs/nuttx_auav-x21_default.cmake [hunk 2]
```diff
@@ -13,7 +13,7 @@ set(config_module_list
 	drivers/magnetometer
 	drivers/telemetry
 
-	drivers/batt_smbus
+	drivers/battery
```
Comment (LorenzMeier) [inline cmake/configs/nuttx_auav-x21_default.cmake]: I'm not sure we should do the rename. After all if I read this correctly we are implementing this standard?
http://sbs-forum.org/specs/sbdat110.pdf

**Comment by LorenzMeier** [comment]
Please cross-check if the part is actually implementing the SBS protocol. If it is, then the name should stay as-is.

File: cmake/configs/nuttx_auav-x21_default.cmake (hunk 2 above)
Comment (AlexKlimaj) [reply cmake/configs/nuttx_auav-x21_default.cmake]: Each fuel gauge will have it's own special additions that warrant an individual driver.

**Comment by AlexKlimaj** [comment]
@dagar We have an rc.battery_setup startup script that updates the battery parameters. I am planning on reworking this driver to inherit from module base. In that refactor, I am going to switch to using parameters and the driver for that functionality and killing the startup script.

File: src/modules/mavlink/mavlink_messages.cpp (hunk 1 above)
Comment (AlexKlimaj) [reply src/modules/mavlink/mavlink_messages.cpp]: The old way just takes the pack voltage and divides by 4.

File: cmake/configs/nuttx_auav-x21_default.cmake (hunk 2 above)
Comment (LorenzMeier) [reply cmake/configs/nuttx_auav-x21_default.cmake]: Why can't those be kept in individual functions?

File: src/modules/mavlink/mavlink_messages.cpp (hunk 1 above)
Comment (LorenzMeier) [reply src/modules/mavlink/mavlink_messages.cpp]: That's not the point. For a non-SMBus equipped drone you're now potentially feeding garbage to the GCS. Have you verified that the individual voltages are all valid for the ADC-driven battery readout?

File: cmake/configs/nuttx_auav-x21_default.cmake (hunk 2 above)
Comment (AlexKlimaj) [reply cmake/configs/nuttx_auav-x21_default.cmake]: It helps with future maintainability and portability.

**Comment by dakejahl** [comment]
@LorenzMeier The BQ40Z50 does indeed implement the SBS protocol. The thought behind this rename is: 
1. The name of the directory conveys the functionality of the files within. e.g: the `battery` directory contains driver code that is associated with the battery
2. The sub directories are device specific. Some fuel gauges might implement the SBS protocol, some may not. Reorganizing the directory structure gives someone the option to easily develop another driver for an arbitrary fuel gauge and stick it in this folder. 

I think naming BQ40Z50 is much more explicit than batt_smbus. This also discourages developers from trying to hack in another device into this generically titled driver.

**Comment by dagar** [comment]
If there's a large portion of this that's effectively a common generic SBS battery driver you pull all of that out into a library.

For example all airspeed (differential pressure) drivers inherit https://github.com/PX4/Firmware/tree/master/src/lib/drivers/airspeed.
https://github.com/PX4/Firmware/tree/master/src/drivers/differential_pressure

**Comment by dakejahl** [comment]
@dagar Yes I agree we should definitely do that, but since there is only a single driver I would like to defer that effort

File: msg/battery_status.msg [hunk 3]
```diff
@@ -18,6 +18,8 @@ uint16 run_time_to_empty   # predicted remaining battery capacity based on the p
 uint16 average_time_to_empty   # predicted remaining battery capacity based on the average rate of discharge in min
 uint16 serial_number       # serial number of the battery pack
 
+float32[4] voltage_cell_v   # Battery individual cell voltages
```
Comment (dagar) [inline msg/battery_status.msg]: This message is getting pretty badly overloaded. It started out as the status of an analog current and voltage sensor.

File: src/drivers/battery/bq40z50/bq40z50.h [hunk 4]
```diff
@@ -116,10 +132,10 @@ int manufacturer_name();
  */
 int serial_number();
 
-extern device::Device *BATT_SMBUS_I2C_interface(int bus);
-typedef device::Device *(*BATT_SMBUS_constructor)(int);
+extern device::Device *BQ40Z50_I2C_interface(int bus);
+typedef device::Device *(*BQ40Z50_constructor)(int);
 
-class BATT_SMBUS : public device::CDev
+class BQ40Z50 : public device::CDev
```
Comment (dagar) [inline src/drivers/battery/bq40z50/bq40z50.h]: This doesn't actually need to be a CDev, although it's fine to keep it this way for now if you want.

File: msg/battery_status.msg (hunk 3 above)
Comment (AlexKlimaj) [reply msg/battery_status.msg]: What would you like to see? Break it out into two messages?

Comment (dagar) [reply msg/battery_status.msg]: It's fine in this PR, but it's yet another sign that the entire battery and system power architecture isn't quite right. We have different fields mashed together for different sources and no way of knowing which fields are expected to be there. For downstream modules to depend on this information (eg commander trigger a battery failsafe) it needs to be treated more like an interface.

Ideas for later - 

We should be able to support a variable number of batteries from any source (analog, smbus, mavlink).

 - battery_status (many)
   - analog (most pixhawks have > 1 now, but only parameters for the first)
   - smbus (should be able to have as many of this as you want)
   - mavlink
 - system_power
   - 5v and 3.3v status
   - usb power valid
   - servo rail voltage
   - number of analog bricks connected

All of this should probably be rolled together into a power module where we can configure what's available on the actual system.

File: src/drivers/battery/bq40z50/bq40z50.cpp [hunk 5]
```diff
@@ -785,7 +785,7 @@ int BQ40Z50::lifetime_read_block_one()
 {
 
 	uint8_t lifetime_block_one_size = 19;
-	uint16_t lifetime_block_one[lifetime_block_one_size] = {0};
```
Comment (dagar) [inline src/drivers/battery/bq40z50/bq40z50.cpp]: The problem is the 0, not the initializer. A single 0 would be for the first field only, not the entire array.

``` C
uint16_t lifetime_block_one[lifetime_block_one_size]{};
```

**Comment by bartslinger** [comment]
I have used this driver for a battery that used another chip than bq40z50. I think it should not be renamed because batt_smbus is actually quite descriptive.

**Comment by dagar** [comment]
@bartslinger would you be able to provide the model so we can figure out how to structure supporting multiple?

https://github.com/PX4/Firmware/pull/10321#issuecomment-415884983

How much of this bq40z50 is still applicable? The register map is probably an okay overview for comparison. https://github.com/PX4/Firmware/pull/10321/files#diff-3de0dd0b616ac3e82862ee3645e51402

**Comment by dagar** [comment]
Quite a lot is from the Smart Battery Data Specification. http://sbs-forum.org/specs/sbdat110.pdf

image

**Comment by bartslinger** [comment]
http://www.ti.com/power-management/battery-management/fuel-gauges/products.html#p199=SMBus
This shows a list of 9 TI chips that use Smart Battery Data Specification.

I've also seen this protocol being used internally in a DJI Phantom 3 battery, but also in a [computer power supply](http://www.fsplifestyle.com/PROP167001465/) and laptop batteries.

**Comment by AlexKlimaj** [comment]
I like the idea about having a batt_smbus base class that can be extended. Then there is a default plain batt_smbus driver that can be started, or any extended versions.

**Pull request closed without merging**